### DISCUSSION TITLE
Deploying in kubernetes with multiple replicas

### DISCUSSION THREAD

**Discussion by adhasahar97** [docker-mailserver/docker-mailserver#3226] [Q&A]
Here is my plan:
1. I'm planning to run this with multiple replica.
2. Mount the all the volumes used by the mailserver and will be shared across all replica. So all of the replica will share the same volume.
3. Create a service that will loadbalance the traffic into each of the replica.

Will I encounter an issue when doing this? Is it straightforward just by increasing the replica number or I need additional configuration for this setup?

Thanks in advance!

**Comment by polarathene** [comment]
> 2\. So all of the replica will share the same volume.

- Assuming NFS for remote storage, I think Dovecot docs mention some compatibility issues there.
- Our change detection service was modified back in v10.x to better support NFS mounts, but AFAIK was a misunderstanding of the issue, and the contributed solution was more of a regression to our locking support.

---

Some users and maintainers use k8s, not sure about multiple replicas (_the prior maintainer / contributor of the changes for NFS was using multiple instances of DMS with shared remote volume though_).

**Comment by dustinmoris** [comment]
Hi, did anyone have any luck running DMS with more than 1 replica? I tried it with an NFS volume and non NFS volume (where the replicas are run on the same node, so no need for a RWX, because RWO is per k8s node) and both times I am unable to get the second replica work properly. The logs keep spamming this on the second replica:

```
mail postfix/master[5535]: fatal: open lock file /var/lib/postfix/master.lock: unable to set exclusive lock: Resource temporarily unavailable
```

Alternatively, if so few people are running DMS with more than 1 replica then how do you ensure high availability?

**Reply by dustinmoris** [reply]
I think in order to make DMS highly available and fault tolerant (who wants to miss important emails or perhaps discover a few weeks late that DMS went down for a reason?) it needs support for shared state, so that multiple replicas can operate on the same volume. Fixing all issues with NFS would be probably the best solution. I don't think it's feasible to have multiple replicas run with their own attached volume behind a load balancer because the incoming emails would be divided up between the different backing instances and depending on where you get connected to you would see a different inbox every time, which is basically akin to a buggy mailserver.

What are your thoughts? Am I missing something?
